File: jp-minimal-android/app/src/main/java/br/com/jpminimal/ui/components/Components.kt
package br.com.jpminimal.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.graphics.toColorInt
import br.com.jpminimal.data.ProductCardDto
import br.com.jpminimal.ui.theme.Gold
import br.com.jpminimal.ui.theme.Ink
import br.com.jpminimal.ui.theme.Line
import br.com.jpminimal.ui.theme.Mist
import br.com.jpminimal.ui.theme.Paper
import br.com.jpminimal.ui.theme.Sand
import br.com.jpminimal.util.formatBRL
import br.com.jpminimal.visual.getProductEmoji

fun parseHex(hex: String): Color = runCatching {
    Color("#${hex.removePrefix("#")}".toColorInt())
}.getOrDefault(Ink)

@Composable
fun ProductVisualBox(
    motif: String,
    colorHex: String,
    modifier: Modifier = Modifier,
) {
    val visual = getProductEmoji(motif)
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(16.dp))
            .background(parseHex(colorHex)),
        contentAlignment = Alignment.Center,
    ) {
        Text(text = visual.emoji, fontSize = (42 * visual.scale).sp)
    }
}

@Composable
fun ProductCard(
    product: ProductCardDto,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(20.dp))
            .background(Color.White.copy(alpha = 0.7f))
            .clickable(onClick = onClick)
            .padding(bottom = 12.dp),
    ) {
        Box(Modifier.fillMaxWidth().aspectRatio(5f / 5.5f)) {
            ProductVisualBox(
                motif = product.visual,
                colorHex = product.colorHex,
                modifier = Modifier.fillMaxSize(),
            )
            Row(
                Modifier.padding(10.dp),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                product.discountPct?.let {
                    BadgeChip("$it% off")
                }
                if (!product.inStock) BadgeChip("Esgotado", filled = false)
            }
        }
        Text(
            product.categoryName.uppercase(),
            style = MaterialTheme.typography.labelSmall,
            color = Mist,
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
        )
        Text(
            product.name,
            style = MaterialTheme.typography.titleMedium,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.padding(horizontal = 12.dp),
        )
        Spacer(Modifier.height(4.dp))
        Row(
            Modifier.padding(horizontal = 12.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(formatBRL(product.price), style = MaterialTheme.typography.bodyMedium)
            product.compareAtPrice?.takeIf { it > product.price }?.let {
                Text(formatBRL(it), color = Mist, style = MaterialTheme.typography.bodySmall)
            }
        }
    }
}

@Composable
fun BadgeChip(text: String, filled: Boolean = true) {
    Text(
        text = text.uppercase(),
        color = if (filled) Paper else Ink,
        fontSize = 10.sp,
        modifier = Modifier
            .clip(RoundedCornerShape(50))
            .background(if (filled) Ink else Paper.copy(alpha = 0.9f))
            .padding(horizontal = 8.dp, vertical = 4.dp),
    )
}

@Composable
fun LoadingBlock(label: String = "Carregando…") {
    Column(
        Modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        CircularProgressIndicator(color = Gold)
        Spacer(Modifier.height(12.dp))
        Text(label, color = Mist)
    }
}

@Composable
fun ErrorBlock(message: String, onRetry: () -> Unit) {
    Column(
        Modifier.fillMaxWidth().padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(message, color = Mist)
        Spacer(Modifier.height(8.dp))
        TextButton(onClick = onRetry) { Text("Tentar de novo", color = Gold) }
    }
}

@Composable
fun JpFilterChip(
    selected: Boolean,
    onClick: () -> Unit,
    label: String,
    modifier: Modifier = Modifier,
) {
    FilterChip(
        selected = selected,
        onClick = onClick,
        modifier = modifier,
        label = {
            Text(
                label,
                maxLines = 1,
                overflow = TextOverflow.Clip,
                softWrap = false,
                color = if (selected) Paper else Ink,
            )
        },
        colors = FilterChipDefaults.filterChipColors(
            containerColor = Paper,
            labelColor = Ink,
            selectedContainerColor = Ink,
            selectedLabelColor = Paper,
        ),
        border = FilterChipDefaults.filterChipBorder(
            enabled = true,
            selected = selected,
            borderColor = Line,
            selectedBorderColor = Ink,
        ),
    )
}

@Composable
fun JpButton(text: String, onClick: () -> Unit, modifier: Modifier = Modifier, enabled: Boolean = true) {
    Button(
        onClick = onClick,
        enabled = enabled,
        modifier = modifier.height(48.dp),
        colors = ButtonDefaults.buttonColors(containerColor = Ink, contentColor = Paper),
        shape = RoundedCornerShape(50),
    ) {
        Text(
            text.uppercase(),
            style = MaterialTheme.typography.labelLarge,
            color = Paper,
            maxLines = 1,
        )
    }
}

@Composable
fun DemoBanner() {
    Box(
        Modifier
            .fillMaxWidth()
            .background(Sand)
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        Text(
            "Loja de portfólio — pagamentos simulados",
            style = MaterialTheme.typography.labelSmall,
            color = Mist,
        )
    }
}

File: jp-minimal-ios/JPMinimal/Components/Components.swift
import SwiftUI
import UIKit

struct JPWordmark: View {
    var body: some View {
        HStack(spacing: 0) {
            Text("JP")
                .font(TypeScale.wordmark)
                .tracking(1)
                .foregroundStyle(Palette.gold)
            Text("  MINIMAL")
                .font(TypeScale.wordmark)
                .tracking(1)
                .foregroundStyle(Palette.ink)
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel("JP Minimal")
    }
}

struct DemoBanner: View {
    var body: some View {
        Text("Loja de portfólio — pagamentos simulados")
            .font(TypeScale.labelSmall)
            .tracking(1.6)
            .foregroundStyle(Palette.mist)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(Palette.sand)
    }
}

struct ProductVisualBox: View {
    var motif: String
    var colorHex: String
    var corner: CGFloat = 16

    var body: some View {
        let visual = ProductEmoji.visual(for: motif)
        GeometryReader { geo in
            ZStack {
                Color(hex: colorHex)
                Text(visual.emoji)
                    .font(.system(size: max(28, min(geo.size.width, geo.size.height) * 0.42 * visual.scale)))
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: corner, style: .continuous))
    }
}

struct BadgeChip: View {
    var text: String
    var filled: Bool = true

    var body: some View {
        Text(text.uppercased())
            .font(.system(size: 10, weight: .semibold))
            .foregroundStyle(filled ? Palette.paper : Palette.ink)
            .padding(.horizontal, 8)
            .padding(.vertical, 4)
            .background(
                Capsule().fill(filled ? Palette.ink : Palette.paper.opacity(0.9))
            )
    }
}

struct ProductCardView: View {
    let product: ProductCard
    var onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: 0) {
                ZStack(alignment: .topLeading) {
                    ProductVisualBox(motif: product.visual, colorHex: product.colorHex)
                        .aspectRatio(1, contentMode: .fit)
                    HStack(spacing: 6) {
                        if let pct = product.discountPct {
                            BadgeChip(text: "\(pct)% off")
                        }
                        if !product.inStock {
                            BadgeChip(text: "Esgotado", filled: false)
                        }
                    }
                    .padding(10)
                }
                Text(product.categoryName.uppercased())
                    .font(TypeScale.labelSmall)
                    .tracking(1.6)
                    .foregroundStyle(Palette.mist)
                    .padding(.horizontal, 12)
                    .padding(.top, 6)
                Text(product.name)
                    .font(TypeScale.titleMedium)
                    .tracking(0.2)
                    .foregroundStyle(Palette.ink)
                    .lineLimit(2)
                    .multilineTextAlignment(.leading)
                    .padding(.horizontal, 12)
                    .padding(.top, 2)
                HStack(spacing: 8) {
                    Text(Money.formatBRL(product.price))
                        .font(TypeScale.bodyMedium)
                        .foregroundStyle(Palette.ink)
                    if let compare = product.compareAtPrice, compare > product.price {
                        Text(Money.formatBRL(compare))
                            .font(TypeScale.bodySmall)
                            .foregroundStyle(Palette.mist)
                            .strikethrough()
                    }
                }
                .padding(.horizontal, 12)
                .padding(.top, 4)
                .padding(.bottom, 12)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color.white.opacity(0.7), in: RoundedRectangle(cornerRadius: 20, style: .continuous))
        }
        .buttonStyle(.plain)
    }
}

struct LoadingBlock: View {
    var label: String = "Carregando…"

    var body: some View {
        VStack(spacing: 12) {
            ProgressView()
                .tint(Palette.gold)
            Text(label)
                .font(TypeScale.bodyMedium)
                .foregroundStyle(Palette.mist)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(32)
    }
}

struct ErrorBlock: View {
    var message: String
    var onRetry: () -> Void

    var body: some View {
        VStack(spacing: 8) {
            Text(message)
                .font(TypeScale.bodyMedium)
                .foregroundStyle(Palette.mist)
                .multilineTextAlignment(.center)
            Button("Tentar de novo", action: onRetry)
                .font(TypeScale.bodyMedium)
                .foregroundStyle(Palette.gold)
                .buttonStyle(.plain)
        }
        .frame(maxWidth: .infinity)
        .padding(24)
    }
}

struct JpFilterChip: View {
    var label: String
    var selected: Bool
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(label)
                .font(TypeScale.bodyMedium)
                .foregroundStyle(selected ? Palette.paper : Palette.ink)
                .lineLimit(1)
                .fixedSize(horizontal: true, vertical: false)
                .padding(.horizontal, 14)
                .padding(.vertical, 8)
                .background(selected ? Palette.ink : Palette.paper, in: Capsule())
                .overlay(Capsule().stroke(selected ? Palette.ink : Palette.line, lineWidth: 1))
        }
        .buttonStyle(.plain)
        .tint(Palette.ink)
    }
}

struct JpButton: View {
    var title: String
    var enabled: Bool = true
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title.uppercased())
                .font(TypeScale.labelLarge)
                .tracking(1.4)
                .foregroundStyle(Palette.paper)
                .frame(maxWidth: .infinity)
                .frame(height: 48)
                .background(enabled ? Palette.ink : Palette.ink.opacity(0.35), in: Capsule())
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .tint(Palette.ink)
    }
}

struct JPField: View {
    var title: String
    @Binding var text: String
    var keyboard: UIKeyboardType = .default
    var isSecure: Bool = false
    var autocapitalization: TextInputAutocapitalization = .sentences

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title)
                .font(TypeScale.labelSmall)
                .tracking(1.6)
                .foregroundStyle(Palette.mist)
            Group {
                if isSecure {
                    SecureField(title, text: $text)
                } else {
                    TextField(title, text: $text)
                        .textInputAutocapitalization(autocapitalization)
                        .keyboardType(keyboard)
                }
            }
            .font(TypeScale.bodyLarge)
            .foregroundStyle(Palette.ink)
            .padding(.horizontal, 14)
            .padding(.vertical, 12)
            .background(Palette.paper)
            .overlay(RoundedRectangle(cornerRadius: 12, style: .continuous).stroke(Palette.line, lineWidth: 1))
            .tint(Palette.gold)
        }
    }
}

struct JPToolbarModifier: ViewModifier {
    var showSearch: Bool = true
    var onSearch: () -> Void

    func body(content: Content) -> some View {
        content
            .navigationBarTitleDisplayMode(.inline)
            .toolbarBackground(Palette.paper, for: .navigationBar)
            .toolbarBackground(.visible, for: .navigationBar)
            .toolbarColorScheme(.light, for: .navigationBar)
            .toolbar {
                ToolbarItem(placement: .principal) { JPWordmark() }
                if showSearch {
                    ToolbarItem(placement: .topBarTrailing) {
                        Button(action: onSearch) {
                            Image(systemName: "magnifyingglass")
                                .foregroundStyle(Palette.ink)
                        }
                        .accessibilityLabel("Buscar")
                    }
                }
            }
    }
}

extension View {
    func jpToolbar(showSearch: Bool = true, onSearch: @escaping () -> Void) -> some View {
        modifier(JPToolbarModifier(showSearch: showSearch, onSearch: onSearch))
    }
}
